### PR TITLE
Issue 7305 : Deploying Pravega with TLS enabled via System test framework

### PR DESCRIPTION
**Pull request by amit-kumar59**
**Change log description**  
Deploying Pravega with TLS enabled via System test framework.

**Purpose of the change**  
Fixes https://github.com/pravega/pravega/issues/7305

**What the code does**  
Made the required changes for TLS enable for the system test.

**How to verify it**  
Run the system test with security and TLS enable system test should be pass.

### CODE REVIEW

File: test/system/src/main/java/io/pravega/test/system/framework/services/kubernetes/K8SequentialExecutor.java [hunk 1]
```diff
@@ -147,9 +151,16 @@ private V1Pod getTestPod(String className, String methodName, String podName) {
                 ))
                 .restartPolicy("Never"));
         if (Utils.TLS_AND_AUTH_ENABLED) {
+            List<URI> ips = Futures.getAndHandleExceptions(client.getStatusOfPodWithLabel(NAMESPACE, "component", "pravega-controller")
```
Comment (anishakj) [inline test/system/src/main/java/io/pravega/test/system/framework/services/kubernetes/K8SequentialExecutor.java:154]: ips is not getting used, due to that no need to pass client to getTestPod function

**Review by anishakj — COMMENTED** [review]
All the changes in controllerresolverfactory can be reverted, as we are not using that

**Comment by amit-kumar59** [comment]
> All the changes in controllerresolverfactory can be reverted, as we are not using that

**Comment by amit-kumar59** [comment]
> All the changes in controllerresolverfactory can be reverted, as we are not using that

Addressed.

File: test/system/src/main/java/io/pravega/test/system/framework/kubernetes/K8sClient.java [hunk 2]
```diff
@@ -670,30 +671,63 @@ public CompletableFuture<V1Status> deleteSecret(String name, String namespace) {
...
-                                        return podStatus.getContainerStatuses()
-                                                .stream()
-                                                .allMatch(st -> st.getState().getRunning() != null);
-                                    }
-                                }).count()),
-                runCount -> { // Number of pods which are running
-                    log.info("Expected running pod count of {}:{}, actual running pod count of {}:{}.", labelValue, expectedPodCount, labelValue, runCount);
-                    if (runCount == expectedPodCount) {
-                        shouldRetry.set(false);
+                        .thenCompose(v -> getPodsWithLabel(namespace, labelName, labelValue)) // fetch status of pods with the given label.
```
Comment (abhinb) [inline test/system/src/main/java/io/pravega/test/system/framework/kubernetes/K8sClient.java:695]: is this change related to tls?

File: test/system/src/test/java/io/pravega/test/system/DynamicRestApiTest.java [hunk 3]
```diff
@@ -140,4 +155,42 @@ public void listScopes() {
...
+    private static class NullHostnameVerifier implements HostnameVerifier {
+        public boolean verify(String hostname, SSLSession session) {
+            return true;
+        }
+    }
+
+    /**
+     * Trust manager that does not perform nay checks.
+     */
+    private static class NullX509TrustManager implements X509TrustManager {
```
Comment (abhinb) [inline test/system/src/test/java/io/pravega/test/system/DynamicRestApiTest.java:183]: maybe we need a TODO here that actually checks the server is trusted with the truststore configured correctly.

File: test/system/src/test/java/io/pravega/test/system/MultiReaderWriterWithFailOverTest.java [hunk 4]
```diff
@@ -92,7 +93,11 @@ public void setup() {
         // Fetch all the RPC endpoints and construct the client URIs.
         final List<String> uris = conURIs.stream().filter(ISGRPC).map(URI::getAuthority).collect(Collectors.toList());
 
-        controllerURIDirect = URI.create("tcp://" + String.join(",", uris));
+        if (Utils.TLS_AND_AUTH_ENABLED) {
```
[the remark this answers is not in the document]
Comment (amit-kumar59) [reply test/system/src/test/java/io/pravega/test/system/MultiReaderWriterWithFailOverTest.java]: Addressed.

File: test/system/src/test/java/io/pravega/test/system/ReadTxnWriteAutoScaleWithFailoverTest.java [hunk 5]
```diff
@@ -55,6 +55,7 @@ public class ReadTxnWriteAutoScaleWithFailoverTest extends AbstractFailoverTests
     private static final int INIT_NUM_WRITERS = 2;
     private static final int ADD_NUM_WRITERS = 2;
     private static final int NUM_READERS = 2;
+    private static final int CONTROLLER_GRPC_PORT = 9090;
```
[the remark this answers is not in the document]
Comment (amit-kumar59) [reply test/system/src/test/java/io/pravega/test/system/ReadTxnWriteAutoScaleWithFailoverTest.java]: Addressed.

File: test/system/src/test/java/io/pravega/test/system/DynamicRestApiTest.java (hunk 3 above)
Comment (amit-kumar59) [reply test/system/src/test/java/io/pravega/test/system/DynamicRestApiTest.java:185]: kept the TODO comments.

File: test/system/src/main/java/io/pravega/test/system/framework/kubernetes/K8sClient.java (hunk 2 above)
Comment (amit-kumar59) [reply test/system/src/main/java/io/pravega/test/system/framework/kubernetes/K8sClient.java:695]: This change is part of https://github.com/pravega/pravega/pull/6445

File: controller/src/main/java/io/pravega/controller/server/ControllerServiceStarter.java [hunk 6]
```diff
@@ -240,11 +240,13 @@ protected void startUp() {
                 SegmentContainerMonitorHealthContributor segmentContainerMonitorHC = new SegmentContainerMonitorHealthContributor("segmentContainerMonitor", monitor );
                 healthServiceManager.register(segmentContainerMonitorHC);
             }
-
+            String controllerServiceName = Config.getConfig("tlsCertCNName", "pravega-pravega-controller");
```
Comment (munish1789) [inline controller/src/main/java/io/pravega/controller/server/ControllerServiceStarter.java:243]: Can we make default to "localhost"
So that if property  tlsCertCNName is not set it will fall back to original case that is localhost.

File: build.gradle [hunk 7]
```diff
@@ -1257,6 +1259,7 @@ project('test:system') {
 
         systemProperty "securityEnabled", System.getProperty("securityEnabled", "false")
         systemProperty "tlsEnabled", System.getProperty("tlsEnabled", "false")
+        systemProperty "tlsCertCNName", System.getProperty("tlsCertCNName", "pravega-pravega-controller")
```
Comment (munish1789) [inline build.gradle:1262]: pravega-pravega-controller can be defaulted to localhost.

File: build.gradle [hunk 8]
```diff
@@ -1156,6 +1156,8 @@ project('test:system') {
         environment "helmRepository", System.getProperty("helmRepository", "https://charts.pravega.io")
         environment "helmHookImageName", System.getProperty("helmHookImageName", "lachlanevenson/k8s-kubectl")
         environment "tlsEnabled", System.getProperty("tlsEnabled", "false")
+        environment "securityEnabled", System.getProperty("securityEnabled", "false")
+        environment "tlsCertCNName", System.getProperty("tlsCertCNName", "pravega-pravega-controller")
```
Comment (munish1789) [inline build.gradle:1160]: pravega-pravega-controller can be defaulted to localhost. and System test can pass this parameter "tlsCertCNName" via jenkins.

File: test/system/src/main/java/io/pravega/test/system/framework/Utils.java [hunk 9]
```diff
@@ -231,11 +238,23 @@ private static boolean isAuthEnabled() {
...
+        return Boolean.valueOf(tlsEnabled) && isAuthEnabled();
     }
 
     public static boolean isSkipLogDownloadEnabled() {
         String config = getConfig("skipLogDownload", "false");
         return config.trim().equalsIgnoreCase("true") ? true : false;
     }
+
+    public static String getTlsCommonName() {
+               return Utils.getConfig("tlsCertCNName", "pravega-pravega-controller");
```
Comment (munish1789) [inline test/system/src/main/java/io/pravega/test/system/framework/Utils.java:250]: here also change default to localhost.

File: controller/src/main/java/io/pravega/controller/server/ControllerServiceStarter.java (hunk 6 above)
Comment (amit-kumar59) [reply controller/src/main/java/io/pravega/controller/server/ControllerServiceStarter.java]: Made default as localhost.

File: build.gradle (hunk 7 above)
Comment (amit-kumar59) [reply build.gradle:1262]: Made default as localhost.

File: build.gradle (hunk 8 above)
Comment (amit-kumar59) [reply build.gradle:1160]: Made default as localhost.

File: test/system/src/main/java/io/pravega/test/system/framework/Utils.java (hunk 9 above)
Comment (amit-kumar59) [reply test/system/src/main/java/io/pravega/test/system/framework/Utils.java]: Made default as localhost.

**Comment by amit-kumar59** [comment]
Run the whole system run with below combinations and its passing.
Tls and Auth Enabled
2816
2818

Auth Enabled
2817

Tls and Auth Disabled
2819

**Review by anishakj — APPROVED** [review]
LGTM

**Review by munish1789 — APPROVED** [review]

**Review by abhinb — APPROVED** [review]
